File: app/core/PhraseCombinatorialEngine.php
<?php
/**
 * Rule-based combinatorial Hiligaynon medical phrase engine (PHP).
 * Matches millions of expressions dynamically from JSON rule tables.
 */

final class PhraseCombinatorialEngine
{
  private const ENGINE_DIR = '/data/nlp/phrase_engine';

  /** @var array<string, mixed>|null */
  private static ?array $symptomRoots = null;

  /** @var list<array<string, mixed>>|null */
  private static ?array $bodyParts = null;

  /** @var array<string, string>|null */
  private static ?array $misspellingMap = null;

  /** @var array{0: array<string, array<string, mixed>>, 1: array<string, array<string, mixed>>, 2: array<string, array<string, mixed>>}|null */
  private static ?array $classificationIndex = null;

  public static function normalize(string $text): string
  {
    $text = strtolower(trim($text));
    $text = preg_replace('/[^a-z0-9\s\-]/', ' ', $text) ?? $text;

    return trim(preg_replace('/\s+/', ' ', $text) ?? $text);
  }

  public static function applyMisspellingCorrections(string $text): string
  {
    $working = self::normalize($text);
    if ($working === '') {
      return '';
    }
    $map = self::misspellingMap();
    uksort($map, static fn (string $a, string $b): int => strlen($b) <=> strlen($a));
    foreach ($map as $wrong => $correct) {
      $working = preg_replace('/(?<!\w)' . preg_quote($wrong, '/') . '(?!\w)/u', $correct, $working) ?? $working;
    }

    return trim($working);
  }

  /** @return list<array<string, mixed>> */
  public static function matchPhrases(string $text): array
  {
    $corrected = self::applyMisspellingCorrections($text);
    $working = self::normalize($corrected);
    if ($working === '') {
      return [];
    }

    $symptomMatches = self::findSymptomMatches($working);
    $bodyMatches = self::findBodyMatches($working);
    $pairs = self::pairSymptomBody($symptomMatches, $bodyMatches);

    $results = [];
    $seen = [];

    foreach ($pairs as [$sMatch, $bMatch]) {
      $root = is_array($sMatch) ? ($sMatch['root'] ?? null) : null;
      $part = is_array($bMatch) ? ($bMatch['part'] ?? null) : null;
      $classification = self::classify(is_array($root) ? $root : null, is_array($part) ? $part : null, $working);
      $english = self::composeEnglish(
        is_array($root) ? $root : null,
        is_array($part) ? $part : null,
        $classification
      );
      $matchedSub = self::extractMatchedSubstring($working, $sMatch, $bMatch);
      $key = $matchedSub . '|' . $english;
      if (isset($seen[$key])) {
        continue;
      }
      $seen[$key] = true;

      $bodyEng = is_array($part) ? (string) ($part['eng'] ?? '') : '';
      $symEng = is_array($root) ? (string) ($root['english_symptom'] ?? '') : '';

      $results[] = [
        'hiligaynon_term'  => $matchedSub !== '' ? $matchedSub : $working,
        'english_term'     => $english,
        'medical_category' => $classification['medical_category'],
        'severity'         => ucfirst($classification['severity']),
        'triage_level'     => $classification['triage_level'],
        'body_part'        => $bodyEng,
        'symptom'          => $symEng,
        'condition'        => (is_array($root) && !empty($root['is_condition'])) || str_contains(strtolower($english), 'infection')
          ? $english : '',
        'source'           => 'phrase_combinatorial_engine',
        'matched_phrase'   => $matchedSub !== '' ? $matchedSub : $working,
        'span'             => self::combinedSpan($sMatch, $bMatch, strlen($working)),
        'confidence'       => 94,
        'engine'           => 'combinatorial_v1',
      ];
    }

    return $results;
  }

  /** @return array<string, int|bool> */
  public static function estimateCombinationCount(): array
  {
    $roots = self::symptomRoots();
    $parts = self::bodyParts();
    $templates = self::loadJson('templates.json');
    $misspellings = self::loadJson('misspelling_rules.json');

    $nSymptomTerms = 0;
    foreach ($roots as $root) {
      $nSymptomTerms += count($root['terms'] ?? []);
    }
    $nBody = count($parts);
    $nOrders = count($templates['word_orders'] ?? []) + count($templates['standalone_orders'] ?? []);
    $nPoss = max(count($templates['possessives'] ?? []), 1);
    $nIntens = max(count($templates['intensifiers'] ?? []), 1);
    $nMisspell = 0;
    foreach (($misspellings['known_variants'] ?? []) as $variants) {
      $nMisspell += is_array($variants) ? count($variants) : 0;
    }

    $symptomBody = $nSymptomTerms * $nBody * max($nOrders, 1) * $nPoss * $nIntens;
    $standalone = $nSymptomTerms * max(count($templates['standalone_orders'] ?? [1]), 1) * $nPoss;
    $withMisspell = ($symptomBody + $standalone) * max((int) ($nMisspell / max($nSymptomTerms, 1)), 1);

    return [
      'symptom_roots'                      => count($roots),
      'symptom_terms'                      => $nSymptomTerms,
      'body_parts'                         => $nBody,
      'templates'                          => $nOrders,
      'theoretical_symptom_body_phrases'   => $symptomBody,
      'theoretical_with_variants'          => $withMisspell,
      'target_met'                         => $withMisspell >= 1_000_000,
    ];
  }

  /** @return list<array<string, mixed>> */
  private static function symptomRoots(): array
  {
    if (self::$symptomRoots !== null) {
      return self::$symptomRoots;
    }
    $data = self::loadJson('symptom_roots.json');
    $roots = [];
    foreach ($data['roots'] ?? [] as $root) {
      $terms = $root['terms'] ?? [];
      usort($terms, static fn (string $a, string $b): int => strlen($b) <=> strlen($a));
      $root['terms_sorted'] = $terms;
      $roots[] = $root;
    }
    usort($roots, static function (array $a, array $b): int {
      $la = max(array_map('strlen', $a['terms_sorted'] ?? ['']));
      $lb = max(array_map('strlen', $b['terms_sorted'] ?? ['']));

      return $lb <=> $la;
    });
    self::$symptomRoots = $roots;

    return self::$symptomRoots;
  }

  /** @return list<array<string, mixed>> */
  private static function bodyParts(): array
  {
    if (self::$bodyParts !== null) {
      return self::$bodyParts;
    }
    $data = self::loadJson('body_parts.json');
    $parts = $data['parts'] ?? [];
    usort($parts, static fn (array $a, array $b): int => strlen((string) ($b['hil'] ?? '')) <=> strlen((string) ($a['hil'] ?? '')));
    self::$bodyParts = $parts;

    return self::$bodyParts;
  }

  /** @return array<string, string> */
  private static function misspellingMap(): array
  {
    if (self::$misspellingMap !== null) {
      return self::$misspellingMap;
    }
    $rules = self::loadJson('misspelling_rules.json');
    $map = [];
    foreach (($rules['known_variants'] ?? []) as $correct => $variants) {
      $c = self::normalize((string) $correct);
      if (!is_array($variants)) {
        continue;
      }
      foreach ($variants as $v) {
        $w = self::normalize((string) $v);
        if ($w !== '' && $w !== $c && !isset($map[$w])) {
          $map[$w] = $c;
        }
      }
    }
    self::$misspellingMap = $map;

    return self::$misspellingMap;
  }

  /** @return array<string, mixed> */
  private static function loadJson(string $name): array
  {
    $path = BASE_PATH . self::ENGINE_DIR . '/' . $name;
    if (!is_readable($path)) {
      return [];
    }
    $raw = file_get_contents($path);
    if ($raw === false) {
      return [];
    }
    $decoded = json_decode($raw, true);

    return is_array($decoded) ? $decoded : [];
  }

  /** @return list<array<string, mixed>> */
  private static function findSymptomMatches(string $text): array
  {
    $matches = [];
    $occupied = array_fill(0, max(strlen($text), 1), false);
    foreach (self::symptomRoots() as $root) {
      foreach ($root['terms_sorted'] ?? [] as $term) {
        $termN = self::normalize((string) $term);
        if ($termN === '') {
          continue;
        }
        if (!preg_match_all('/(?<!\w)' . preg_quote($termN, '/') . '(?!\w)/u', $text, $m, PREG_OFFSET_CAPTURE)) {
          continue;
        }
        foreach ($m[0] as [$_, $start]) {
          $end = $start + strlen($termN);
          $overlap = false;
          for ($i = $start; $i < $end; $i++) {
            if ($occupied[$i] ?? false) {
              $overlap = true;
              break;
            }
          }
          if ($overlap) {
            continue;
          }
          for ($i = $start; $i < $end; $i++) {
            $occupied[$i] = true;
          }
          $matches[] = ['root' => $root, 'term' => $termN, 'span' => [$start, $end]];
        }
      }
    }
    usort($matches, static fn (array $a, array $b): int => ($a['span'][0] ?? 0) <=> ($b['span'][0] ?? 0));

    return $matches;
  }

  /** @return list<array<string, mixed>> */
  private static function findBodyMatches(string $text): array
  {
    $matches = [];
    $occupied = array_fill(0, max(strlen($text), 1), false);
    foreach (self::bodyParts() as $part) {
      $hil = self::normalize((string) ($part['hil'] ?? ''));
      if ($hil === '') {
        continue;
      }
      if (!preg_match_all('/(?<!\w)' . preg_quote($hil, '/') . '(?!\w)/u', $text, $m, PREG_OFFSET_CAPTURE)) {
        continue;
      }
      foreach ($m[0] as [$_, $start]) {
        $end = $start + strlen($hil);
        $overlap = false;
        for ($i = $start; $i < $end; $i++) {
          if ($occupied[$i] ?? false) {
            $overlap = true;
            break;
          }
        }
        if ($overlap) {
          continue;
        }
        for ($i = $start; $i < $end; $i++) {
          $occupied[$i] = true;
        }
        $matches[] = ['part' => $part, 'term' => $hil, 'span' => [$start, $end]];
      }
    }
    usort($matches, static fn (array $a, array $b): int => ($a['span'][0] ?? 0) <=> ($b['span'][0] ?? 0));

    return $matches;
  }

  /**
   * @param list<array<string, mixed>> $symptomMatches
   * @param list<array<string, mixed>> $bodyMatches
   * @return list<array{0: array<string, mixed>|null, 1: array<string, mixed>|null}>
   */
  private static function pairSymptomBody(array $symptomMatches, array $bodyMatches): array
  {
    if ($symptomMatches === [] && $bodyMatches === []) {
      return [];
    }
    if ($symptomMatches === []) {
      return array_map(static fn (array $b): array => [null, $b], $bodyMatches);
    }
    if ($bodyMatches === []) {
      return array_map(static fn (array $s): array => [$s, null], $symptomMatches);
    }

    $pairs = [];
    $usedBodies = [];
    foreach ($symptomMatches as $s) {
      $bestBody = null;
      $bestIdx = -1;
      $bestDist = 999;
      foreach ($bodyMatches as $idx => $b) {
        if (isset($usedBodies[$idx])) {
          continue;
        }
        $dist = abs((int) (($s['span'][0] + $s['span'][1]) / 2) - (int) (($b['span'][0] + $b['span'][1]) / 2));
        if ($dist < $bestDist) {
          $bestDist = $dist;
          $bestBody = $b;
          $bestIdx = $idx;
        }
      }
      if ($bestBody !== null && $bestDist <= 40) {
        $usedBodies[$bestIdx] = true;
        $pairs[] = [$s, $bestBody];
      } else {
        $pairs[] = [$s, null];
      }
    }
    foreach ($bodyMatches as $idx => $b) {
      if (!isset($usedBodies[$idx])) {
        $pairs[] = [null, $b];
      }
    }

    return $pairs;
  }

  /**
   * @param array<string, mixed>|null $root
   * @param array<string, mixed>|null $part
   * @return array{triage_level: string, severity: string, medical_category: string, english_override: string}
   */
  private static function classify(?array $root, ?array $part, string $text): array
  {
    [$pairRules, $defaults, $patterns] = self::classificationIndex();

    foreach ($patterns as $pat => $rule) {
      if (str_contains($text, (string) $pat)) {
        return [
          'triage_level'     => (string) ($rule['triage'] ?? 'emergency'),
          'severity'         => (string) ($rule['severity'] ?? 'critical'),
          'medical_category' => (string) ($rule['category'] ?? 'general'),
          'english_override' => (string) ($rule['english'] ?? ''),
        ];
      }
    }

    $symptomId = (string) ($root['id'] ?? '');
    $bodyEng = (string) ($part['eng'] ?? '');

    if ($symptomId !== '' && $bodyEng !== '' && isset($pairRules[$symptomId . '|' . $bodyEng])) {
      $rule = $pairRules[$symptomId . '|' . $bodyEng];

      return [
        'triage_level'     => (string) ($rule['triage'] ?? 'urgent'),
        'severity'         => (string) ($rule['severity'] ?? 'moderate'),
        'medical_category' => (string) ($rule['category'] ?? 'general'),
        'english_override' => '',
      ];
    }

    if ($symptomId !== '' && isset($defaults[$symptomId])) {
      $d = $defaults[$symptomId];

      return [
        'triage_level'     => (string) ($d['triage'] ?? 'urgent'),
        'severity'         => (string) ($d['severity'] ?? 'moderate'),
        'medical_category' => (string) ($root['category'] ?? 'general'),
        'english_override' => '',
      ];
    }

    if ($root !== null) {
      return [
        'triage_level'     => (string) ($root['default_triage'] ?? 'routine'),
        'severity'         => (string) ($root['default_severity'] ?? 'moderate'),
        'medical_category' => (string) ($root['category'] ?? 'general'),
        'english_override' => '',
      ];
    }

    return [
      'triage_level'     => 'routine',
      'severity'         => 'mild',
      'medical_category' => (string) ($part['category'] ?? 'general'),
      'english_override' => '',
    ];
  }

  /**
   * @param array<string, mixed>|null $root
   * @param array<string, mixed>|null $part
   * @param array{triage_level: string, severity: string, medical_category: string, english_override: string} $classification
   */
  private static function composeEnglish(?array $root, ?array $part, array $classification): string
  {
    if ($classification['english_override'] !== '') {
      return $classification['english_override'];
    }

    if ($root !== null && $part !== null) {
      $engSym = (string) ($root['english_symptom'] ?? 'symptom');
      $engBody = (string) ($part['eng'] ?? '');
      if (($root['id'] ?? '') === 'pus_infection') {
        return $engBody . ' infection';
      }
      if (in_array($engSym, ['pain', 'swelling', 'itching', 'redness', 'bleeding'], true)) {
        return $engBody . ' ' . $engSym;
      }

      return $engBody . ' ' . (string) ($root['english_phrase'] ?? $engSym);
    }

    if ($root !== null) {
      return (string) (($root['english_phrase'] ?? '') ?: ($root['english_symptom'] ?? 'symptom'));
    }

    if ($part !== null) {
      return (string) ($part['eng'] ?? 'body part');
    }

    return 'symptom';
  }

  /** @param array<string, mixed>|null $sMatch @param array<string, mixed>|null $bMatch */
  private static function extractMatchedSubstring(string $text, ?array $sMatch, ?array $bMatch): string
  {
    $spans = [];
    if (is_array($sMatch)) {
      $spans[] = $sMatch['span'];
    }
    if (is_array($bMatch)) {
      $spans[] = $bMatch['span'];
    }
    if ($spans === []) {
      return $text;
    }
    $start = min(array_column($spans, 0));
    $end = max(array_column($spans, 1));

    return trim(substr($text, $start, $end - $start));
  }

  /** @param array<string, mixed>|null $sMatch @param array<string, mixed>|null $bMatch @return list<int> */
  private static function combinedSpan(?array $sMatch, ?array $bMatch, int $len): array
  {
    $starts = [];
    $ends = [];
    if (is_array($sMatch)) {
      $starts[] = $sMatch['span'][0];
      $ends[] = $sMatch['span'][1];
    }
    if (is_array($bMatch)) {
      $starts[] = $bMatch['span'][0];
      $ends[] = $bMatch['span'][1];
    }
    if ($starts === []) {
      return [0, $len];
    }

    return [min($starts), max($ends)];
  }

  /** @return array{0: array<string, array<string, mixed>>, 1: array<string, array<string, mixed>>, 2: array<string, array<string, mixed>>} */
  private static function classificationIndex(): array
  {
    if (self::$classificationIndex !== null) {
      return self::$classificationIndex;
    }
    $data = self::loadJson('classification_rules.json');
    $pairRules = [];
    foreach ($data['rules'] ?? [] as $rule) {
      $key = ($rule['symptom_id'] ?? '') . '|' . ($rule['body_eng'] ?? '');
      $pairRules[$key] = $rule;
    }
    $defaults = is_array($data['defaults_by_symptom'] ?? null) ? $data['defaults_by_symptom'] : [];
    $patterns = [];
    foreach ($data['pattern_rules'] ?? [] as $rule) {
      if (!empty($rule['pattern'])) {
        $patterns[(string) $rule['pattern']] = $rule;
      }
    }
    self::$classificationIndex = [$pairRules, $defaults, $patterns];

    return self::$classificationIndex;
  }
}

File: ai_service/phrase_combinatorial_engine.py
"""
Rule-based combinatorial Hiligaynon medical phrase engine.

Generates and matches millions of symptom expressions dynamically from:
  symptom roots + body parts + templates + misspelling rules + classification rules

Does NOT require millions of static CSV rows.
"""

from __future__ import annotations

import json
import re
from functools import lru_cache
from pathlib import Path
from typing import Any

_ENGINE_DIR = Path(__file__).resolve().parent.parent / "data" / "nlp" / "phrase_engine"


def normalize(text: str) -> str:
    text = (text or "").lower().strip()
    text = re.sub(r"[^a-z0-9\s\-]", " ", text)
    return re.sub(r"\s+", " ", text).strip()


@lru_cache(maxsize=1)
def _load_json(name: str) -> dict[str, Any]:
    path = _ENGINE_DIR / name
    if not path.is_file():
        return {}
    with path.open(encoding="utf-8") as f:
        return json.load(f)


@lru_cache(maxsize=1)
def symptom_roots() -> tuple[dict[str, Any], ...]:
    data = _load_json("symptom_roots.json")
    roots: list[dict[str, Any]] = []
    for root in data.get("roots", []):
        terms = sorted((root.get("terms") or []), key=len, reverse=True)
        roots.append({**root, "terms_sorted": terms})
    roots.sort(key=lambda r: max((len(t) for t in r.get("terms_sorted", [])), default=0), reverse=True)
    return tuple(roots)


@lru_cache(maxsize=1)
def body_parts() -> tuple[dict[str, Any], ...]:
    data = _load_json("body_parts.json")
    parts = list(data.get("parts", []))
    parts.sort(key=lambda p: len(p.get("hil", "")), reverse=True)
    return tuple(parts)


@lru_cache(maxsize=1)
def misspelling_map() -> dict[str, str]:
    """wrong -> correct term (longest keys first at apply time)."""
    rules = _load_json("misspelling_rules.json")
    mapping: dict[str, str] = {}
    for correct, variants in (rules.get("known_variants") or {}).items():
        c = normalize(correct)
        for v in variants:
            w = normalize(v)
            if w and w != c and w not in mapping:
                mapping[w] = c
    return mapping


@lru_cache(maxsize=1)
def classification_index() -> tuple[dict[tuple[str, str], dict[str, Any]], dict[str, dict[str, Any]], dict[str, dict[str, Any]]]:
    data = _load_json("classification_rules.json")
    pair_rules: dict[tuple[str, str], dict[str, Any]] = {}
    for rule in data.get("rules", []):
        key = (rule.get("symptom_id", ""), rule.get("body_eng", ""))
        pair_rules[key] = rule
    defaults = data.get("defaults_by_symptom") or {}
    patterns = {r["pattern"]: r for r in data.get("pattern_rules", []) if r.get("pattern")}
    return pair_rules, defaults, patterns


def apply_misspelling_corrections(text: str) -> str:
    working = normalize(text)
    if not working:
        return ""
    mp = misspelling_map()
    for wrong in sorted(mp.keys(), key=len, reverse=True):
        correct = mp[wrong]
        working = re.sub(r"(?<!\w)" + re.escape(wrong) + r"(?!\w)", correct, working)
    return working.strip()


def _find_symptom_matches(text: str) -> list[dict[str, Any]]:
    matches: list[dict[str, Any]] = []
    occupied = [False] * max(len(text), 1)
    for root in symptom_roots():
        for term in root.get("terms_sorted", []):
            term_n = normalize(term)
            if not term_n:
                continue
            pattern = re.compile(r"(?<!\w)" + re.escape(term_n) + r"(?!\w)")
            for m in pattern.finditer(text):
                start, end = m.start(), m.end()
                if any(occupied[start:end]):
                    continue
                for i in range(start, end):
                    occupied[i] = True
                matches.append({
                    "root": root,
                    "term": term_n,
                    "span": [start, end],
                })
    matches.sort(key=lambda x: x["span"][0])
    return matches


def _find_body_matches(text: str) -> list[dict[str, Any]]:
    matches: list[dict[str, Any]] = []
    occupied = [False] * max(len(text), 1)
    for part in body_parts():
        hil = normalize(part.get("hil", ""))
        if not hil:
            continue
        pattern = re.compile(r"(?<!\w)" + re.escape(hil) + r"(?!\w)")
        for m in pattern.finditer(text):
            start, end = m.start(), m.end()
            if any(occupied[start:end]):
                continue
            for i in range(start, end):
                occupied[i] = True
            matches.append({
                "part": part,
                "term": hil,
                "span": [start, end],
            })
    matches.sort(key=lambda x: x["span"][0])
    return matches


def _pair_symptom_body(
    symptom_matches: list[dict[str, Any]],
    body_matches: list[dict[str, Any]],
) -> list[tuple[dict[str, Any] | None, dict[str, Any] | None]]:
    if not symptom_matches and not body_matches:
        return []
    if not symptom_matches:
        return [(None, b) for b in body_matches]
    if not body_matches:
        return [(s, None) for s in symptom_matches]

    pairs: list[tuple[dict[str, Any] | None, dict[str, Any] | None]] = []
    used_bodies: set[int] = set()
    for s in symptom_matches:
        best_body = None
        best_dist = 999
        for idx, b in enumerate(body_matches):
            if idx in used_bodies:
                continue
            dist = abs((s["span"][0] + s["span"][1]) // 2 - (b["span"][0] + b["span"][1]) // 2)
            if dist < best_dist:
                best_dist = dist
                best_body = (idx, b)
        if best_body and best_dist <= 40:
            used_bodies.add(best_body[0])
            pairs.append((s, best_body[1]))
        else:
            pairs.append((s, None))
    for idx, b in enumerate(body_matches):
        if idx not in used_bodies:
            pairs.append((None, b))
    return pairs


def _classify(
    root: dict[str, Any] | None,
    part: dict[str, Any] | None,
    text: str,
) -> dict[str, str]:
    pair_rules, defaults, patterns = classification_index()
    for pat, rule in patterns.items():
        if pat in text:
            return {
                "triage_level": rule.get("triage", "emergency"),
                "severity": rule.get("severity", "critical"),
                "medical_category": rule.get("category", "general"),
                "english_override": rule.get("english", ""),
            }

    symptom_id = (root or {}).get("id", "")
    body_eng = (part or {}).get("eng", "")

    if symptom_id and body_eng:
        rule = pair_rules.get((symptom_id, body_eng))
        if rule:
            return {
                "triage_level": rule.get("triage", "urgent"),
                "severity": rule.get("severity", "moderate"),
                "medical_category": rule.get("category", "general"),
                "english_override": "",
            }

    if symptom_id in defaults:
        d = defaults[symptom_id]
        return {
            "triage_level": d.get("triage", "urgent"),
            "severity": d.get("severity", "moderate"),
            "medical_category": (root or {}).get("category", "general"),
            "english_override": "",
        }

    if root:
        return {
            "triage_level": root.get("default_triage", "routine"),
            "severity": root.get("default_severity", "moderate"),
            "medical_category": root.get("category", "general"),
            "english_override": "",
        }

    return {
        "triage_level": "routine",
        "severity": "mild",
        "medical_category": (part or {}).get("category", "general"),
        "english_override": "",
    }


def _compose_english(
    root: dict[str, Any] | None,
    part: dict[str, Any] | None,
    classification: dict[str, str],
) -> str:
    if classification.get("english_override"):
        return classification["english_override"]

    if root and part:
        eng_sym = root.get("english_symptom", "symptom")
        eng_body = part.get("eng", "")
        if root.get("id") == "pus_infection":
            return f"{eng_body} infection"
        if eng_sym in ("pain", "swelling", "itching", "redness", "bleeding"):
            return f"{eng_body} {eng_sym}"
        return f"{eng_body} {root.get('english_phrase', eng_sym)}"

    if root:
        return str(root.get("english_phrase") or root.get("english_symptom", "symptom"))

    if part:
        return str(part.get("eng", "body part"))

    return "symptom"


def _extract_matched_substring(text: str, symptom: dict[str, Any] | None, body: dict[str, Any] | None) -> str:
    spans: list[tuple[int, int]] = []
    if symptom:
        spans.append(tuple(symptom["span"]))
    if body:
        spans.append(tuple(body["span"]))
    if not spans:
        return text
    start = min(s[0] for s in spans)
    end = max(s[1] for s in spans)
    return text[start:end].strip()


def match_phrases(text: str) -> list[dict[str, Any]]:
    """Dynamically match combinatorial symptom phrases in patient text."""
    corrected = apply_misspelling_corrections(text)
    working = normalize(corrected)
    if not working:
        return []

    symptom_matches = _find_symptom_matches(working)
    body_matches = _find_body_matches(working)
    pairs = _pair_symptom_body(symptom_matches, body_matches)

    results: list[dict[str, Any]] = []
    seen: set[str] = set()

    for s_match, b_match in pairs:
        root = s_match["root"] if s_match else None
        part = b_match["part"] if b_match else None
        classification = _classify(root, part, working)
        english = _compose_english(root, part, classification)
        matched_sub = _extract_matched_substring(working, s_match, b_match)
        key = f"{matched_sub}|{english}"
        if key in seen:
            continue
        seen.add(key)

        body_eng = (part or {}).get("eng", "")
        sym_eng = (root or {}).get("english_symptom", "")

        results.append({
            "hiligaynon_term": matched_sub or working,
            "english_term": english,
            "medical_category": classification["medical_category"],
            "severity": classification["severity"].capitalize() if classification["severity"] else "Moderate",
            "triage_level": classification["triage_level"],
            "body_part": body_eng,
            "symptom": sym_eng,
            "condition": english if (root or {}).get("is_condition") or "infection" in english.lower() else "",
            "source": "phrase_combinatorial_engine",
            "matched_phrase": matched_sub or working,
            "span": [
                min(
                    s_match["span"][0] if s_match else len(working),
                    b_match["span"][0] if b_match else len(working),
                ),
                max(
                    s_match["span"][1] if s_match else 0,
                    b_match["span"][1] if b_match else 0,
                ),
            ],
            "confidence": 94,
            "engine": "combinatorial_v1",
        })

    return results


def estimate_combination_count() -> dict[str, int]:
    """Theoretical coverage from rule tables (not stored rows)."""
    roots = symptom_roots()
    parts = body_parts()
    templates = _load_json("templates.json")
    misspellings = _load_json("misspelling_rules.json")

    n_symptom_terms = sum(len(r.get("terms", [])) for r in roots)
    n_body = len(parts)
    n_orders = len(templates.get("word_orders", [])) + len(templates.get("standalone_orders", []))
    n_poss = max(len(templates.get("possessives", [])), 1)
    n_intens = max(len(templates.get("intensifiers", [])), 1)
    n_misspell = sum(len(v) for v in (misspellings.get("known_variants") or {}).values())

    symptom_body = n_symptom_terms * n_body * n_orders * n_poss * n_intens
    standalone = n_symptom_terms * len(templates.get("standalone_orders", [1])) * n_poss
    with_misspell = (symptom_body + standalone) * max(n_misspell // max(n_symptom_terms, 1), 1)

    return {
        "symptom_roots": len(roots),
        "symptom_terms": n_symptom_terms,
        "body_parts": n_body,
        "templates": n_orders,
        "theoretical_symptom_body_phrases": symptom_body,
        "theoretical_with_variants": with_misspell,
        "target_met": with_misspell >= 1_000_000,
    }


def generate_canonical_phrase(symptom_term: str, body_hil: str, possessive: str = "ko") -> str:
    """Generate one canonical phrase (for exemplar export, not bulk storage)."""
    return normalize(f"{symptom_term} {body_hil} {possessive}")


def clear_cache() -> None:
    _load_json.cache_clear()
    symptom_roots.cache_clear()
    body_parts.cache_clear()
    misspelling_map.cache_clear()
    classification_index.cache_clear()
